impl/src/main/java/org/glassfish/soteria/authorization/spi/impl/SubjectParser.java

```diff
@@ -389,181 +383,200 @@ private boolean tryWebLogic(String contextID, Collection<String> allDeclaredRole
-                    return principal;
-                // JBoss EAP/WildFly convention 2 - the one and only principal in group called CallerPrincipal
-                case "org.jboss.security.SimpleGroup": 
-                    if (principal.getName().equals("CallerPrincipal") && principal instanceof Group) {
-                        
-                        Enumeration<? extends Principal> groupMembers = ((Group) principal).members();
-                        
-                        if (groupMembers.hasMoreElements()) {
-                            return groupMembers.nextElement();
-                        }
-                    }
-                    break;
-                case "org.apache.tomee.catalina.TomcatSecurityService$TomcatUser": // TomEE
-                    try {
-                        return 
-                            (Principal) Class.forName("org.apache.catalina.realm.GenericPrincipal")
-                                             .getMethod("getUserPrincipal")
-                                             .invoke(
-                                                 Class.forName("org.apache.tomee.catalina.TomcatSecurityService$TomcatUser")
-                                                      .getMethod("getTomcatPrincipal")
-                                                      .invoke(principal))
-                            
-                            ;
-                        } catch (Exception e) {
-                            
-                        }
-                    break;
-            }
-
+            // If the principal is a standard one, return it
             if (CallerPrincipal.class.isAssignableFrom(principal.getClass())) {
                 return principal;
             }
-            
+
+            // Do some checks to determine it from vendor specific data
+            Principal vendorCallerPrincipal = getVendorCallerPrincipal(principal);
+            if (vendorCallerPrincipal != null) {
+                return vendorCallerPrincipal;
+            }
+
+            // Check for Servlet
             try {
                 return CDI.current().select(HttpServletRequest.class).get().getUserPrincipal();
             } catch (Exception e) {
                 // Not inside an HttpServletRequest
             }
-         
+
+            // Check for EJB
             EJBContext ejbContext = EJB.getEJBContext();
             if (ejbContext != null) {
-                String principalName = ejbContext.getCallerPrincipal().getName();
-             
-                // EJBs return empty/null name for unauthenticated callers
-                return principalName != null && !principalName.isEmpty() ? ejbContext.getCallerPrincipal() : null;
+                // The EJB returned value must be verified for its "authenticated name" since it's vendor specific
+                return getVendorCallerPrincipal(ejbContext.getCallerPrincipal());
```

Comment:
The logic of `getVendorCallerPrincipal()` would be simpler if the check for EJB "unauthenticated" principal took place only in this context, rather than being pushed down into `getVendorCallerPrincipal()`, which now must check whether it's in EJB context on every invocation, rather than just the calls originating from this line of code. Alternatively, an "isEjb" flag could be passed as a parameter, to avoid the lookup.